File: 2024-11-November-GFG-POTD/November 22 - Stock Buy and Sell – Max one Transaction Allowed/Solution.cpp
// User function Template for C++

#include <iostream>
#include <vector>
#include <climits> // For INT_MAX
using namespace std;

class Solution {
public:
    int maximumProfit(vector<int>& prices) {
        int minPrice = INT_MAX;  // Tracks the minimum price encountered
        int maxProfit = 0;       // Tracks the maximum profit found
        
        // Traverse the prices array
        for (int price : prices) {
            if (price < minPrice) {
                minPrice = price;  // Update minimum price
            } else {
                // Calculate profit and update maxProfit if current profit is greater
                maxProfit = max(maxProfit, price - minPrice);
            }
        }
        return maxProfit; // Return maximum profit (0 if no profitable transaction)
    }
};

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n; // Number of prices
        vector<int> prices(n);
        for (int i = 0; i < n; i++) {
            cin >> prices[i];
        }
        Solution obj;
        int res = obj.maximumProfit(prices);
        cout << res << endl;
    }
    return 0;
}

File: 2024-11-November-GFG-POTD/November 22 - Stock Buy and Sell – Max one Transaction Allowed/Solution.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.lang.*;
import java.util.*;

class GFG {
    public static void main(String args[]) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            String arr[] = br.readLine().split(" ");
            int prices[] = new int[arr.length];

            for (int i = 0; i < arr.length; i++) {
                prices[i] = Integer.parseInt(arr[i]);
            }
            Solution obj = new Solution();
            int res = obj.maximumProfit(prices);
            System.out.println(res);
        }
    }
}
// } Driver Code Ends


// User function Template for Java

class Solution {
    public int maximumProfit(int prices[]) {
        // Code here
        // Initialize variables
        int minPrice = Integer.MAX_VALUE; // Tracks minimum price encountered
        int maxProfit = 0;                // Tracks maximum profit found
        
        // Traverse the prices array
        for (int price : prices) {
            if (price < minPrice) {
                minPrice = price; // Update minimum price
            } else {
                // Calculate profit and update maxProfit if current profit is greater
                maxProfit = Math.max(maxProfit, price - minPrice);
            }
        }
        return maxProfit; // Return maximum profit (0 if no profitable transaction)
    }
}
